vita3k/app/src/app_init.cpp

```diff
@@ -101,11 +102,86 @@ void update_viewport(EmuEnvState &state) {
     }
 }
 
+void init_paths(Root &root_paths) {
+    const auto dir_sep = std::string{ fs::path::preferred_separator };
+    auto base_path = SDL_GetBasePath();
+    auto pref_path = SDL_GetPrefPath(org_name, app_name);
+    root_paths.set_base_path(string_utils::utf_to_wide(base_path));
+    root_paths.set_pref_path(string_utils::utf_to_wide(pref_path));
+    root_paths.set_log_path(string_utils::utf_to_wide(base_path));
+    root_paths.set_config_path(string_utils::utf_to_wide(base_path));
+    root_paths.set_shared_path(string_utils::utf_to_wide(base_path));
+    root_paths.set_cache_path(fs::path(string_utils::utf_to_wide(base_path)) / "cache" / dir_sep);
+    SDL_free(base_path);
+    SDL_free(pref_path);
+
+#if defined(__linux__) && !defined(__ANDROID__) && !defined(__APPLE__)
+    // XDG Data Dirs.
+    auto env_home = getenv("HOME");
+    auto env_data = getenv("XDG_DATA_HOME");
+    auto env_home_data = env_data;
+    if (env_data != NULL)
+        root_paths.set_pref_path(fs::path(string_utils::utf_to_wide(env_data)) / app_name / app_name / dir_sep);
+
+    env_data = getenv("XDG_CONFIG_HOME");
+    if (env_data != NULL)
+        root_paths.set_config_path(fs::path(string_utils::utf_to_wide(env_data)) / app_name / dir_sep);
+    else if (env_home != NULL)
+        root_paths.set_config_path(fs::path(env_home) / ".config" / app_name / dir_sep);
+
+    env_data = getenv("XDG_CACHE_HOME");
+    if (env_data != NULL) {
+        root_paths.set_cache_path(fs::path(string_utils::utf_to_wide(env_data)) / app_name / dir_sep);
+        root_paths.set_log_path(fs::path(string_utils::utf_to_wide(env_data)) / app_name / dir_sep);
+    } else if (env_home != NULL) {
+        root_paths.set_cache_path(fs::path(env_home) / ".cache" / app_name / dir_sep);
+        root_paths.set_log_path(fs::path(env_home) / ".cache" / app_name / dir_sep);
+    }
+
+    // Don't assume that base_path is portable.
+    if (fs::exists(root_paths.get_base_path() / "data") && fs::exists(root_paths.get_base_path() / "lang") && fs::exists(root_paths.get_base_path() / "shaders-builtin"))
+        root_paths.set_shared_path(root_paths.get_base_path());
+    else if (env_home != NULL)
+        root_paths.set_shared_path(fs::path(env_home) / ".local/share" / app_name / dir_sep);
+
+    env_data = getenv("XDG_DATA_DIRS");
+    if (env_data != NULL) {
+        auto env_paths = string_utils::split_string(env_data, ':');
+        for (std::vector<std::string>::iterator iter = env_paths.begin(); iter != env_paths.end(); iter++) {
```

Comment:
would be cleaner to use the `for (auto& i : env_paths)` loop, its shorter and nicer